p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,388 @@
+	msg.ObsAddrs = addrsToBytes(hs.ids.OwnObservedAddrs())
+	tstart := time.Now()
+	if err := wr.WriteMsg(msg); err != nil {
+		s.Reset()
+		hs.handlerError(rp, fmt.Errorf("failed to write CONNECT message to initator:: %w", err))
+		return
+	}
+
+	// Read SYNC message
+	msg.Reset()
+	if err := rd.ReadMsg(msg); err != nil {
+		s.Reset()
+		hs.handlerError(rp, fmt.Errorf("failed to read message from initator: %w", err))
+		return
+	}
+	rtt := time.Since(tstart)
+
+	if t := msg.GetType(); t != pb.HolePunch_SYNC {
+		s.Reset()
+		hs.handlerError(rp, fmt.Errorf("expected SYNC message from initiator but got %d", t))
+		return
+	}
+	defer s.Close()
+
+	// Hole punch now by forcing a connect
+	pi := peer.AddrInfo{
+		ID:    rp,
+		Addrs: obsDial,
+	}
+	hs.tracer.StartHolePunch(rp, obsDial, rtt)
+	log.Debugf("starting hole punch with %s", rp)
+	counter := atomic.AddInt32(&hs.counter, 1)
+	tstart = time.Now()
+	err := hs.holePunchConnect(pi, int(counter))
+	dt := time.Since(tstart)
+	hs.tracer.EndHolePunch(rp, dt, err)
+	if err != nil {
+		log.Warnf("hole punching with %s failed after %s: %s", rp, dt, err)
+	} else {
+		log.Infof("hole punching with %s successful after %s", rp, dt)
+	}
+}
+
+func (hs *Service) holePunchConnect(pi peer.AddrInfo, attempt int) error {
+	holePunchCtx := network.WithSimultaneousConnect(hs.ctx, "hole-punching")
+	forceDirectConnCtx := network.WithForceDirectDial(holePunchCtx, "hole-punching")
+	dialCtx, cancel := context.WithTimeout(forceDirectConnCtx, dialTimeout)
+	defer cancel()
+
+	hs.tracer.HolePunchAttempt(pi.ID, attempt)
+	err := hs.host.Connect(dialCtx, pi)
+	if err == nil {
+		log.Infof("hole punch with peer %s successful after %d retries; direct conns to peer are:", pi.ID, attempt)
+		for _, c := range hs.host.Network().ConnsToPeer(pi.ID) {
+			if !isRelayAddress(c.RemoteMultiaddr()) {
+				log.Info(c)
+			}
+		}
+	}
+	log.Infof("hole punch attempt %d with peer %s failed: %s; will retry in %s...", attempt, pi.ID, err, retryWait)
+	return err
+}
+
+func isRelayAddress(a ma.Multiaddr) bool {
+	_, err := a.ValueForProtocol(ma.P_CIRCUIT)
+	return err == nil
+}
+
+func addrsToBytes(as []ma.Multiaddr) [][]byte {
+	bzs := make([][]byte, 0, len(as))
+	for _, a := range as {
+		bzs = append(bzs, a.Bytes())
+	}
+	return bzs
+}
+
+func addrsFromBytes(bzs [][]byte) []ma.Multiaddr {
+	addrs := make([]ma.Multiaddr, 0, len(bzs))
+	for _, bz := range bzs {
+		a, err := ma.NewMultiaddrBytes(bz)
+		if err == nil {
+			addrs = append(addrs, a)
+		}
+	}
+	return addrs
+}
+
+type netNotifiee Service
+
+func (nn *netNotifiee) HolePunchService() *Service {
+	return (*Service)(nn)
+}
+
+func (nn *netNotifiee) Connected(_ network.Network, v network.Conn) {
+	hs := nn.HolePunchService()
+	dir := v.Stat().Direction
+
+	// Hole punch if it's an inbound proxy connection.
+	// If we already have a direct connection with the remote peer, this will be a no-op.
+	if dir == network.DirInbound && isRelayAddress(v.RemoteMultiaddr()) {
```

Comment:
We should _probably_ bail early if we're _known_ to be dialable. This would largely mitigate a lot of DoS concerns.

Also, is there some discussion on picking a side to initiate? On one hand, having the receiver initiate gives them a chance to try a "normal" dial. On the other hand, it means the receiver may do a _lot_ of work on behalf of a potentially malicious party.

Reply:
> Also, is there some discussion on picking a side to initiate? On one hand, having the receiver initiate gives them a chance to try a "normal" dial. On the other hand, it means the receiver may do a lot of work on behalf of a potentially malicious party.

The node behind the NAT is the one dialing. The reason for that is that there's a good change that the peer is not behind a NAT, and we can just dial him directly. Only if the peer is not dialable, we have to fall back to hole punching.

Reply:
> We should probably bail early if we're known to be dialable. This would largely mitigate a lot of DoS concerns.

Then the remote address wouldn't be a relay address, would it?

Reply:
To clarify, we're looking at the remote address being a relay address, AND the connection being an inbound connection.
If we know to be publicly dialable, we shouldn't be behind a relay in the first place.

Reply:
> The node behind the NAT is the one dialing. The reason for that is that there's a good change that the peer is not behind a NAT, and we can just dial him directly. Only if the peer is not dialable, we have to fall back to hole punching.

Makes sense.

>  If we know to be publicly dialable, we shouldn't be behind a relay in the first place.

That doesn't _stop_ someone from making a relayed connection. We'll accept those regardless (there are some valid use-cases for this, e.g., no common transport).

To know if we're publicly dialable, we need to check our "nat" status (subscribe to the reachability event).

Reply:
That makes sense. Let's do that in a separate PR (this PR is targeting a feature-branch anyway).